docs/AUTH_PROTOCOL.md

```diff
@@ -0,0 +1,393 @@
+# Proposal for alternative design for JSON messages
+
+This is a proposal for a new JSON-based authentication protocol for GDM (and, eventually, the systemd upstream replacement for GDM and other desktop-specific display managers). The authenticator 
+
+This protocol is event driven. Basically, the two parties (the authenticator, which I'll be calling the "broker". And the frontend, which I'll be calling "GDM") will pass events back and forth at each other as things happen. Events have a basic structure:
+
+```js
+{
+    "event": [some event],
+    [arguments]
+}
+```
+
+## PAM
+
+PAM is not event-driven, but instead it acts as a blocking conversation: request, response. Next request, next response. We also need to consider a future beyond PAM. So, this protocol happens entirely outside of the PAM context. Here's how we retrofit the protocol into PAM.
+
+Basically, we'd spin up a specific PAM service and wait for the first question it sends us. That first question might be a normal PAM question, at which point we fall back to legacy PAM behavior. It might be a switchable auth JSON request (i.e. org.gnome.DisplayManager.UserVerifier.CustomJSON, "auth-mechanisms" v1), at which point we go down the GDM switchable auth route (GNOME 50). Or it might be an "upgrade-to-fd" request (e.g. org.gnome.DisplayManager.UserVerifier.SwitchToFd), at which point we do the following:
```

Comment:
what his "gdm switchable auth"?

Reply:
> what his "gdm switchable auth"?

It's a protocol that uses [JSON over PAM through binary conversations](https://gitlab.gnome.org/GNOME/gdm/-/blob/main/pam-extensions/gdm-custom-json-pam-extension.h?ref_type=heads) that allows to do data-only communication between a PAM module and a PAM application, we defined couple of protocols around it at the moment, but we're trying to centralize everything here to an unique protocol.

> also what do you mean by "switch to fd"?
> 
> that comunication contiues via a json-based protocol?

Basically once the communication is kicked-off through systemd, the rest of the communication is private between the authentication service and the UI